File: Linked List Cycle/Linked List Cycle.java
/**
 * Definition for singly-linked list.
 * class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode(int x) {
 *         val = x;
 *         next = null;
 *     }
 * }
 */
public class Solution {
    public boolean hasCycle(ListNode head) {
        
        if(head == null || head.next == null){
            return false;
        }
        
        ListNode fastPtr = head.next;
        
        while(fastPtr != null){
            if(fastPtr == head){
                return true;
            }
            else{
                head = head.next;
                fastPtr = fastPtr.next;
                if(fastPtr == null){
                    return false;
                }
                else{
                    fastPtr = fastPtr.next;
                }
            }
        }
        
        return false;
    }
}

File: Linked List Cycle/Linked List Cycle.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode(int x) : val(x), next(NULL) {}
 * };
 */
 
#define EXTRA 0
 
class Solution {
public:
    bool hasCycle(ListNode *head) {

// Use extra space
#ifdef EXTRA
        std::set<ListNode*> visited;
        
        while(head){
            if(visited.find(head) == visited.end()){
                visited.insert(head);
                head = head->next;
            }
            else
                return true;
        }
        
        return false;
        
// Use NO extra space
#else

        if(!head)   return false;
        ListNode* turtle = head;
        if(!turtle->next)   return false;
        ListNode* rabbit = turtle->next;
        
        while(!turtle && !rabbit){
            if(turtle == rabbit)    return true;
            else{
                turtle = turtle->next;
                rabbit = rabbit->next->next;
            }
        }
        
        return false;

#endif

    }
};
